File: bank-service/src/main/java/rs/edu/raf/service/impl/ExchangeAccountServiceImpl.java
package rs.edu.raf.service.impl;

import lombok.AllArgsConstructor;
import org.springframework.stereotype.Service;
import rs.edu.raf.exceptions.ExchangeAccountNotFoundException;
import rs.edu.raf.model.ExchangeAccount;
import rs.edu.raf.repository.ExchangeAccountRepository;
import rs.edu.raf.service.ExchangeAccountService;
import rs.edu.raf.service.ExchangeRateService;

import java.math.BigDecimal;

@Service
@AllArgsConstructor
public class ExchangeAccountServiceImpl implements ExchangeAccountService {
    private ExchangeAccountRepository exchangeAccountRepository;
    private ExchangeRateService exchangeRateService;


    @Override
    public BigDecimal exchange(String fromCurrency, String toCurrency, BigDecimal amount) throws Exception {
        ExchangeAccount senderExchangeAccount = exchangeAccountRepository.findByCurrencyCode(fromCurrency)
                .orElseThrow(()->new ExchangeAccountNotFoundException(fromCurrency));

        ExchangeAccount receiverExchangeAccount = exchangeAccountRepository.findByCurrencyCode(toCurrency)
                .orElseThrow(()->new ExchangeAccountNotFoundException(toCurrency));

        BigDecimal amountToReceive = exchangeRateService.convertToCurrency(fromCurrency, toCurrency, amount);

        senderExchangeAccount.setBalance(senderExchangeAccount.getBalance().add(amount));
        receiverExchangeAccount.setBalance(receiverExchangeAccount.getBalance().subtract(amountToReceive));
        exchangeAccountRepository.save(senderExchangeAccount);
        exchangeAccountRepository.save(receiverExchangeAccount);

        return amountToReceive;
    }
}

File: bank-service-kotlin/src/main/kotlin/org/example/bankservicekotlin/service/impl/ExchangeAccountServiceImpl.kt
package org.example.bankservicekotlin.service.impl

import org.example.bankservicekotlin.exception.ExchangeAccountNotFoundException
import org.example.bankservicekotlin.model.ExchangeAccount
import org.example.bankservicekotlin.repository.ExchangeAccountRepository
import org.example.bankservicekotlin.service.ExchangeAccountService
import org.example.bankservicekotlin.service.ExchangeRateService
import org.springframework.stereotype.Service
import java.math.BigDecimal

@Service
class ExchangeAccountServiceImpl(
    private val exchangeAccountRepository: ExchangeAccountRepository,
    private val exchangeRateService: ExchangeRateService): ExchangeAccountService {

    @Throws(Exception::class)
    override fun exchange(fromCurrency: String, toCurrency: String, amount: BigDecimal): BigDecimal {
        val senderExchangeAccount: ExchangeAccount = exchangeAccountRepository.findByCurrencyCode(fromCurrency)
            .orElseThrow { ExchangeAccountNotFoundException(fromCurrency) }

        val receiverExchangeAccount: ExchangeAccount = exchangeAccountRepository.findByCurrencyCode(toCurrency)
            .orElseThrow { ExchangeAccountNotFoundException(toCurrency) }

        val amountToReceive = exchangeRateService.convertToCurrency(fromCurrency, toCurrency, amount)

        senderExchangeAccount.balance = senderExchangeAccount.balance.add(amount)
        receiverExchangeAccount.balance = receiverExchangeAccount.balance.subtract(amountToReceive)
        exchangeAccountRepository.save(senderExchangeAccount)
        exchangeAccountRepository.save(receiverExchangeAccount)

        return amountToReceive
    }

}